File: src/main/java/com/challenge/interview/calendar/model/TimeInterval.java
package com.challenge.interview.calendar.model;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.time.OffsetDateTime;

@Getter
@AllArgsConstructor
public class TimeInterval {

    private OffsetDateTime from;
    private OffsetDateTime to;

    public TimeInterval overlap(TimeInterval timeInterval) {
        if (timeInterval.getFrom().isAfter(this.to)
                || this.from.isAfter(timeInterval.getTo())
                || this.to.isBefore(this.from)
                || timeInterval.getTo().isBefore(timeInterval.getFrom())) {
            return null;
        }
        OffsetDateTime intersectedFrom = this.from.isBefore(timeInterval.getFrom()) ? timeInterval.from : this.from;
        OffsetDateTime intersectedTo = this.to.isBefore(timeInterval.to) ?  this.to : timeInterval.to;
        return new TimeInterval(intersectedFrom, intersectedTo);
    }
}

File: src/main/kotlin/com/challenge/interview/calendar/model/TimeInterval.kt
package com.challenge.interview.calendar.model

import java.time.OffsetDateTime

class TimeInterval(val from: OffsetDateTime, val to: OffsetDateTime) {
    fun overlap(timeInterval: TimeInterval): TimeInterval? {
        if (timeInterval.from > this.to
            || this.from > timeInterval.to
            || this.to < this.from
            || timeInterval.to < timeInterval.from) {
            return null
        }
        val intersectedFrom = if (this.from < timeInterval.from) timeInterval.from else this.from
        val intersectedTo = if (this.to < timeInterval.to) this.to else timeInterval.to
        return TimeInterval(intersectedFrom, intersectedTo)
    }
}